Support SCC/PCC drivers

ccli8 opened the pull request:
This PR tries to support SCC/PCC drivers, including:
1. Support system clock controller driver and integrate with `z_arm_platform_init()`
1. Support peripheral clock controller driver and integrate with GPIO/UART drivers.

ccli8 committed:
drivers: clock_control: numaker: Initial support

1.  Support system clock controller driver
2.  Support peripheral clock controller driver

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
soc: numaker: m46x: Integrate with SCC driver

System clock control logic is run at:
1.  z_arm_platform_init() on SCC driver disabled
2.  SCC driver init function on SCC driver enabled

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: uart: numaker: Integrate with PCC driver

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: gpio: numaker: Integrate with PCC driver

NOTE: GPIO port clock enable is moved to GPIO port init function.
      So that its control relies on GPIO port node in devicetree.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

cyliangtw commented on soc/arm/nuvoton_numaker/m46x/soc.c:
```diff
@@ -6,46 +6,87 @@
     /* Init System Clock                                                                                       */
     /*---------------------------------------------------------------------------------------------------------*/
 
-    /* Enable External XTAL (4~24 MHz) */
```
This comment should be reserved

cyliangtw commented on soc/arm/nuvoton_numaker/m46x/soc.c:
```diff
@@ -6,46 +6,87 @@
 
-    /* Enable External XTAL (4~24 MHz) */
-    CLK_EnableXtalRC(CLK_PWRCTL_HXTEN_Msk);
-    /* Waiting for 12MHz clock ready */
```
This comment should be reserved

cyliangtw commented on soc/arm/nuvoton_numaker/m46x/soc.c:
```diff
@@ -6,46 +6,87 @@
+    }
+#endif
 
-    /* Enable HIRC clock (Internal RC 22.1184MHz) */
```
This comment should be reserved

ccli8 replied:
Added back comment, including all above

cyliangtw reviewed (changes requested):
It's not easy to maintain & not clear for simple clock init segment.

cyliangtw commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -226,6 +227,35 @@ static void gpio_numaker_isr(const struct device *dev)
 	gpio_fire_callbacks(&data->callbacks, dev, int_status);
 }
 
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_PCC
+#define NUMAKER_DT_INST_CLOCK_INIT(n)                           \
+    .clk_dev    = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(n)),        \
+    .clk_modidx = DT_INST_CLOCKS_CELL(n, clock_module_index),
+#else
+#define NUMAKER_DT_INST_CLOCK_INIT(n)                           \
+    .clk_modidx = DT_INST_CLOCKS_CELL(n, clock_module_index),
+#endif
+
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_PCC
+#define GPIO_NUMAKER_CLOCK_INIT(n)                  \
+    do {                                            \
+        const struct gpio_numaker_config *config = dev->config;                 \
+        int err = clock_control_on(config->clk_dev, \
+                               (clock_control_subsys_t) config->clk_modidx);    \
+        if (err != 0) {                             \
+            return err;                             \
+        }                                           \
+    } while (0)
+#else
+#define GPIO_NUMAKER_CLOCK_INIT(n)                  \
+    do {                                            \
+        const struct gpio_numaker_config *config = dev->config;                 \
+        SYS_UnlockReg();                            \
+        CLK_EnableModuleClock(config->clk_modidx);  \
+        SYS_LockReg();                              \
+    } while (0)
+#endif
```
I not prefer this way.
It's too complex for simple CLK init procedure.
To store index/info in config structure, let gpio_numaker_configure() handle it instead of macro, like as 
#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_PCC

#else

#endif

ccli8 committed:
drivers: clock_control: numaker: Always enable HIRC/LIRC

Make HIRC/LIRC non-configurable and always enable them

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
boards: arm: numaker_pfm_m467: Disable SCC driver by default

With SCC driver disabled, z_arm_platform_init() will take charge of system clock initialization sequence instead.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented:
Closed via #2
